File: CPP/No82.cc
/**
 * Created by Xiaozhong on 2020/7/29.
 * Copyright (c) 2020/7/29 Xiaozhong. All rights reserved.
 */

#include <iostream>
#include <vector>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;

    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    ListNode *deleteDuplicates(ListNode *head) {
        if (head == nullptr) return nullptr;        // 判断特殊情况
        ListNode *m_ptr = new ListNode(0);          // 负责修改的结点指针
        ListNode *new_head = m_ptr;                 // 记录待返回来的头结点指针
        m_ptr->next = head;
        ListNode *s_ptr = head;
        while (s_ptr != nullptr && s_ptr->next != nullptr) {
            if (s_ptr->val == s_ptr->next->val) {
                int val = s_ptr->val;
                // 一直扫描到节点数值不同为止
                while (s_ptr && s_ptr->val == val) {
                    s_ptr = s_ptr->next;
                }
                // 更新节点过来
                m_ptr->next = s_ptr;
            } else {
                // 没有扫描到一样的节点，那就一起往后走
                m_ptr = m_ptr->next;
                s_ptr = s_ptr->next;
            }
        }
        return new_head->next;
    }
};

int main() {
    ListNode *l1 = new ListNode(1);
    ListNode *l2 = new ListNode(2);
    ListNode *l3 = new ListNode(3);
    ListNode *l4 = new ListNode(3);
    ListNode *l5 = new ListNode(4);
    ListNode *l6 = new ListNode(4);
    ListNode *l7 = new ListNode(5);
    l1->next = l2;
    l2->next = l3;
    l3->next = l4;
    l4->next = l5;
    l5->next = l6;
    l6->next = l7;
    Solution s;
    ListNode *head = s.deleteDuplicates(l1);
    while (head != nullptr) {
        cout << head->val << " ";
        head = head->next;
    }
}

File: Java/src/No82.java
/*
 * Copyright (c) 2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

public class No82 {
    static class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode deleteDuplicates(ListNode head) {
        // 首先检查特殊情况，排除掉
        if (head == null) return null;
        // 设置用来时刻跟进修改进度的指针，
        // 该指针上一次和下一次指向的节点一定中间没有重复值
        ListNode modifyNode = new ListNode(0);
        // 用来帮助找到返回值的哨兵指针，虚拟头节点
        ListNode guard = modifyNode;
        guard.next = head;
        // 用来逐个遍历链表中元素的指针
        ListNode current = head;
        // 逐个检查，前提是当前元素不为空而且下个元素也不为空
        while (current != null && current.next != null) {
            // 如果当前元素和下一个元素值重复了，那么就开始 【跳】，直接让 current 跳到不重复的元素那里
            if (current.val == current.next.val) {
                int val = current.val;
                while (current != null && current.val == val) current = current.next;
                // 然后让跟进修改进度的指针下一个位置指向 current，中间略过重复元素
                modifyNode.next = current;
            } else {
                // 如果不重复，那么这俩指针一起往后走
                current = current.next;
                modifyNode = modifyNode.next;
            }
        }
        // 返回哨兵指针指向的头节点
        return guard.next;
    }
}
